File: project_euler_unified/problem_532/solution.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

/*
 * Problem 532: Nanoscale Strips
 *
 * f(W,H) = C(W+H, W)^2 mod p,  p = 999999937.
 *
 * Precompute factorials and inverse factorials mod p, then evaluate.
 * Time: O(W+H).  Space: O(W+H).
 */

const ll MOD = 999999937LL;
const int MAXN = 20000001;

ll fact[MAXN], inv_fact[MAXN];

ll power(ll base, ll exp, ll mod) {
    ll result = 1;
    base %= mod;
    while (exp > 0) {
        if (exp & 1) result = result * base % mod;
        base = base * base % mod;
        exp >>= 1;
    }
    return result;
}

void precompute(int n) {
    fact[0] = 1;
    for (int i = 1; i <= n; i++)
        fact[i] = fact[i - 1] * i % MOD;
    inv_fact[n] = power(fact[n], MOD - 2, MOD);
    for (int i = n - 1; i >= 0; i--)
        inv_fact[i] = inv_fact[i + 1] * (i + 1) % MOD;
}

ll binom(int n, int k) {
    if (k < 0 || k > n) return 0;
    return fact[n] % MOD * inv_fact[k] % MOD * inv_fact[n - k] % MOD;
}

int main() {
    int W = 10000000, H = 10000000;
    precompute(W + H);
    ll c = binom(W + H, W);
    ll ans = c * c % MOD;
    cout << ans << endl;
    return 0;
}

File: project_euler_unified/problem_532/solution.py
"""
Problem 532: Nanoscale Strips

Compute f(10^7, 10^7) mod 999999937, where f(W,H) = C(W+H, W)^2 mod p.

Method: precompute factorials and inverse factorials mod p, then
evaluate the binomial coefficient in O(1).
"""

MOD = 999999937

def solve(W, H, mod):
    n = W + H
    # Precompute factorials
    fact = [1] * (n + 1)
    for i in range(1, n + 1):
        fact[i] = fact[i - 1] * i % mod
    # Inverse factorials via Fermat's little theorem
    inv_fact = [1] * (n + 1)
    inv_fact[n] = pow(fact[n], mod - 2, mod)
    for i in range(n - 1, -1, -1):
        inv_fact[i] = inv_fact[i + 1] * (i + 1) % mod
    # C(n, W) mod p
    c = fact[n] * inv_fact[W] % mod * inv_fact[H] % mod
    return c * c % mod

print(solve(10**7, 10**7, MOD))
